### PR TITLE
ARROW-15611: [C++] Migrate arrow::ipc::internal::json::ArrayFromJSON to Result<>

### PR DESCRIPTION
**Pull request by AlvinJ15**
Migrate arrow::ipc::internal::json::ArrayFromJSON to Result<>

### CODE REVIEW

File: cpp/src/arrow/python/gdb.cc [hunk 1]
```diff
@@ -29,6 +29,7 @@
 #include "arrow/scalar.h"
 #include "arrow/table.h"
 #include "arrow/type.h"
+#include "arrow/testing/gtest_util.h"
```
Comment (lidavidm) [inline cpp/src/arrow/python/gdb.cc:32]: I'm not sure we want this linked to googletest, we should stick to the old version.

**Review by lidavidm — COMMENTED** [review]
There are build errors that need fixing.

File: cpp/src/arrow/python/gdb.h [hunk 2]
```diff
@@ -23,7 +23,7 @@ namespace arrow {
 namespace gdb {
 
 ARROW_PYTHON_EXPORT
-void TestSession();
+Status TestSession();
```
Comment (lidavidm) [inline cpp/src/arrow/python/gdb.h:26]: We need to update the binding signature on the Cython side.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Er, whoops. Thanks for the changes, there's just a minor fix we should make. I think the S3 test issues are due to ARROW-16043, and we need to figure out why the GDB tests fail.

File: cpp/src/arrow/python/gdb.cc [hunk 3]
```diff
@@ -98,13 +96,13 @@ std::shared_ptr<Array> SliceArrayFromJSON(const std::shared_ptr<DataType>& ty,
...
 
-void TestSession() {
+Status TestSession() {
   // We define local variables for all types for which we want to test
   // pretty-printing.
   // Then, at the end of this function, we trap to the debugger, so that
   // test instrumentation can print values from this frame by interacting
   // with the debugger.
-  // The test instrumentation is in pyarrow/tests/test_gdb.py
+  // The test instrumentation is in pyarrow/tests/test_gdb.pytest_gdb.py
```
Comment (pitrou) [inline cpp/src/arrow/python/gdb.cc:105]: I think that change is not deliberate, is it?

File: cpp/src/arrow/python/gdb.cc [hunk 4]
```diff
@@ -82,13 +82,11 @@ class UuidType : public ExtensionType {
...
-std::shared_ptr<Array> SliceArrayFromJSON(const std::shared_ptr<DataType>& ty,
-                                          util::string_view json, int64_t offset = 0,
-                                          int64_t length = -1) {
-  std::shared_ptr<Array> array;
-  ARROW_CHECK_OK(ArrayFromJSON(ty, json, &array));
+Result<std::shared_ptr<Array>> SliceArrayFromJSON(const std::shared_ptr<DataType>& ty,
+                                                  util::string_view json,
+                                                  int64_t offset = 0,
+                                                  int64_t length = -1) {
+  ARROW_ASSIGN_OR_RAISE(auto array, ArrayFromJSON(ty, json));
```
Comment (pitrou) [inline cpp/src/arrow/python/gdb.cc:89]: I don't think this change is necessary, instead, you could just have:
```suggestion
std::shared_ptr<Array> SliceArrayFromJSON(const std::shared_ptr<DataType>& ty,
                                          util::string_view json, int64_t offset = 0,
                                          int64_t length = -1) {
  auto array = *ArrayFromJSON(ty, json);
```

File: cpp/src/arrow/python/gdb.cc [hunk 5]
```diff
@@ -98,13 +96,13 @@ std::shared_ptr<Array> SliceArrayFromJSON(const std::shared_ptr<DataType>& ty,
 
 }  // namespace
 
-void TestSession() {
+Status TestSession() {
```
Comment (pitrou) [inline cpp/src/arrow/python/gdb.cc:99]: Similarly, this change doesn't seem necessary.

**Review by pitrou — COMMENTED** [review]
Thanks a lot for this, @AlvinJ15 . I think we can simplify things a bit, see comments.

**Review by pitrou — APPROVED** [review]
+1, thanks for the update @AlvinJ15 !

**Pull request closed without merging**